### PR TITLE
[JENKINS-59283] - Use different icons for computers that are temporarily offline

### PR DESCRIPTION
**Pull request by olivergondza**
See [JENKINS-59283](https://issues.jenkins-ci.org/browse/JENKINS-59283).

### Proposed changelog entries

* Use different computer icon for temporary offline state.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by jvz — APPROVED** [review]
LGTM. I suppose it's probably a separate issue, but do we really need the PNGs as well anymore? See https://caniuse.com/#feat=svg for evidence that SVG is widely supported.

**Comment by olivergondza** [comment]
@jvz, they are order of magnitude larger than 48px png icons and the potential to reduce the number of http calls is somewhat limited (we can reuse cached version for different sizes). What I have been tentatively considering is inlining images using [data URIs](https://html.com/attributes/img-src/#Data_URIs) to reduce the number of consecutive HTTP calls to make while making the responses heavier.

**Comment by jvz** [comment]
Ah, ok, makes sense. Possibly less of an issue with HTTP/2, though file size would still be a concern.

**Comment by zbynek** [comment]
>they are order of magnitude larger than 48px png icons
In this PR the 48px icon is 2.89KB, the SVG is 2.86KB when optimized (default settings in https://jakearchibald.github.io/svgomg/ + precision=2 + "sort attrs") and gzipped. Obviously it's not a blocker for the PR, just an idea for future simplification.

**Comment by olivergondza** [comment]
@zbynek, posibly, though are those svgs still "editable"? With all objects, paths, groups, etc. clearly separated so images can be mixed/reused? </ot>

**Comment by zbynek** [comment]
@olivergondza I think some of the information is lost. But you still only need one generated + one source SVG and the result looks good for all device pixel ratios.

File: core/src/main/java/hudson/model/Computer.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Computer.java]: Maybe an off-topic, but using longer self-describing file names could be helpful

**Review by oleg-nenashev — APPROVED** [review]
Nice improvement!

**Comment by jvz** [comment]
> @zbynek, posibly, though are those svgs still "editable"? With all objects, paths, groups, etc. clearly separated so images can be mixed/reused?

Something like this should be part of the maven build. I like the idea of having the source SVG be the one with layers and comments and whatever else make it easy to edit while compiling it to an optimized SVG similar to what's done for other web content nowadays.

**Comment by oleg-nenashev** [comment]
@jvz @zbynek Maybe we could just enable https://imgbot.net/ for starters

**Comment by oleg-nenashev** [comment]
Just set up an example: https://github.com/oleg-nenashev/role-strategy-plugin/pull/1

**Comment by oleg-nenashev** [comment]
And https://github.com/oleg-nenashev/jenkins/pull/38 for the core. Looks like it optimizes only a subset of images in a batch, but looks fine as well

**Comment by olivergondza** [comment]
> And [oleg-nenashev#38](https://github.com/oleg-nenashev/jenkins/pull/38) for the core. Looks like it optimizes only a subset of images in a batch, but looks fine as well

I does not touch the source SVGs and that is good.

**Comment by olivergondza** [comment]
Any more comments w.r.t. the actual change here?

**Comment by oleg-nenashev** [comment]
I am +1. Since @zbynek is not blocking it, I intend merging it tomorrow unless there is negative feedback or request to put it on hold

**Comment by oleg-nenashev** [comment]
Happy to merge it if the matching logic is adjusted

**Comment by ianwilliams1** [comment]
We found the Node to be displayed as on-line when:
`Jenkins.instance.getNode('Node-Name').getComputer().setAcceptingTasks(false)
`
There was no icon differentiator but the node was not accepting tasks; effectively off-line.

Do the changes in this PR cover this scenario? If not, should PR this be revised or an issue separate to [JENKINS-59283](https://issues.jenkins-ci.org/browse/JENKINS-59283) be created?

**Comment by olivergondza** [comment]
@ianwilliams1, it is not. My main goal was to differentiate *someone* has put the node offline. Though I guess this is a good chance to think this through.

**Comment by ianwilliams1** [comment]
> @ianwilliams1, it is not. My main goal was to differentiate _someone_ has put the node offline . Though I guess this is a good chance to think this through.

@olivergondza, I had a really difficult time figuring out what was wrong and how we got into that state as the indicator showed on-line. It had to do w/switching from a scheduled availability to always online but in the wrong window. There is also very poor documentation on the "state" of of  node, eg:

- Difference between temporarily off-line vs Disconnected.
- When to use Availability: "Keep online as much as possible" vs "online when in demand, offline when idle"
- and, whether a given node (vs entire system) accepting tasks or not.

I'm sure most users are similarly confused. Should a separate Issue be created to address this? Several issues?

**Comment by olivergondza** [comment]
I doubt icons are going to fix that. All those states are here for historical reasons and they are not going anywhere so here we can only discuss how to communicate the implications of individual states.

**Comment by ianwilliams1** [comment]
@olivergondza, not suggesting to change any behaviours. just as you suggest, better communication.

So,
- one issue to better document the use cases (Offline vs Disconnected, Online on-demand), 
- one issue to provide better status visual indicators, perhaps with separate indicators for the node status and the node queue status?

**Comment by olivergondza** [comment]
Let agree on the icons to use in individual states first. How about

- Nodes temporarily offline by a human will be yellow (icon introduced by this PR).
- Nodes with no channel, not accepting tasks or temporarily offline for other reason (ping timeout, full disk, etc.) will be the red one.
- All other states will be the default icon.

```java
// The machine was taken offline by someone
if (isTemporarilyOffline() && getOfflineCause() instanceof OfflineCause.UserCause) return "computer-u.png";
// There is a "technical" reason the computer will not accept new builds
if (isOffline() || !isAcceptingTasks()) return "computer-x.png";
return "computer.png";
```

**Comment by oleg-nenashev** [comment]
Fine with me. Later we can have another icon for node monitors

**Comment by timja** [comment]
Does this still need to be on-hold?

**Comment by oleg-nenashev** [comment]
I think so. We have a consensus in comments, but it has not been implemented yet

**Comment by olivergondza** [comment]
Updated to reflect the reviews. (Thanks!)

- Icons have more descriptive names
- Nodes not accepting tasks are displayed the same way as offline
- The "yellow" icon is only used for "user offline" status.

**Review by timja — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
Looks ready to go, +1 for merging.
Since it was around for a week after the recent changes, I will go ahead tonight so that it gets into the weekly.

Thanks for the change @olivergondza !

**Pull request merged**
